webapp/src/components/common/GroupedDataTable/index.tsx

```diff
@@ -1,69 +1,299 @@
+import useEnqueueErrorSnackbar from "../../../hooks/useEnqueueErrorSnackbar";
+import { toError } from "../../../utils/fnUtils";
+import useOperationInProgressCount from "../../../hooks/useOperationInProgressCount";
+import type { TRow } from "./types";
 
-export interface GroupedDataTableProps<TData extends TRow> {
+export interface GroupedDataTableProps<
+  TGroups extends string[],
+  TData extends TRow<TGroups[number]>,
+> {
   data: TData[];
-  columns: Array<MRT_ColumnDef<TData>>;
-  groups: string[] | readonly string[];
-  onCreate?: (values: TData) => Promise<TData>;
-  onDelete?: (ids: string[]) => void;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  columns: Array<MRT_ColumnDef<TData, any>>;
+  groups: TGroups;
+  onCreate?: (values: TRow<TGroups[number]>) => Promise<TData>;
+  onDuplicate?: (row: TData, newName: string) => Promise<TData>;
+  onDelete?: (rows: TData[]) => PromiseAny | void;
+  onNameClick?: (row: TData) => void;
+  isLoading?: boolean;
+  deleteConfirmationMessage?: string | ((count: number) => string);
+  fillPendingRow?: (
+    pendingRow: TRow<TGroups[number]>,
+  ) => TRow<TGroups[number]> & Partial<TData>;
 }
 
-function GroupedDataTable<TData extends TRow>({
+// Use ids to identify default columns (instead of `accessorKey`),
+// to have a unique identifier. It is more likely to have a duplicate
+// `accessorKey` with `columns` prop.
+const GROUP_COLUMN_ID = "_group";
+const NAME_COLUMN_ID = "_name";
+
+function GroupedDataTable<
+  TGroups extends string[],
+  TData extends TRow<TGroups[number]>,
+>({
   data,
   columns,
   groups,
   onCreate,
+  onDuplicate,
   onDelete,
-}: GroupedDataTableProps<TData>) {
+  onNameClick,
+  isLoading,
+  deleteConfirmationMessage,
+  fillPendingRow,
+}: GroupedDataTableProps<TGroups, TData>) {
   const { t } = useTranslation();
   const [openDialog, setOpenDialog] = useState<
     "add" | "duplicate" | "delete" | ""
   >("");
   const [tableData, setTableData] = useState(data);
   const [rowSelection, setRowSelection] = useState<MRT_RowSelectionState>({});
+  const enqueueErrorSnackbar = useEnqueueErrorSnackbar();
+  const callbacksRef = useAutoUpdateRef({ onNameClick });
+  const prevData = usePrevious(data);
+  const pendingRows = useRef<Array<TRow<TGroups[number]>>>([]);
+  const { createOps, deleteOps, totalOps } = useOperationInProgressCount();
 
-  const isAnyRowSelected = useMemo(
-    () => Object.values(rowSelection).some((value) => value),
-    [rowSelection],
-  );
-
-  const isOneRowSelected = useMemo(
-    () => Object.values(rowSelection).filter((value) => value).length === 1,
-    [rowSelection],
-  );
-
-  const selectedRow = useMemo(() => {
-    if (isOneRowSelected) {
-      const selectedIndex = Object.keys(rowSelection).find(
-        (key) => rowSelection[key],
-      );
-      return selectedIndex && tableData[+selectedIndex];
+  // Update once `data` only if previous value was empty.
+  // It allows to handle loading data.
```

Comment:
Yes I know lol, but I use this format only for JSDoc or for top file comment (IDE color is not the same for both). It is very common to see single line comment for long sentences like this.

Reply:
I agree that multi-line comments should be avoided inside code blocks to simplify commenting out sections as needed.But for documenting functions, hooks or constants at the start of blocks (like you did in this case), multi-line comments are suitable as they're typically used for more detailed explanations. Many well-regarded libraries also follow this practice. If your preference for single-line comments is just personal, that's completely fine I'm not advocating for overly strict rules just aiming for clarity and consistency where it might enhance understanding.

Reply:
My initial comment was maybe too short I could've explain the rationale behind